File: ig-codegen-cs/IgCodegen/IgPackageScanner.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace IgCodegen;

/// <summary>
/// Scans a restored FHIR package directory and classifies its resources into
/// CodeSystem/Profile/Extension canonical URL constants.
/// </summary>
public sealed class IgPackageScanner
{
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        AllowTrailingCommas = true,
        ReadCommentHandling = JsonCommentHandling.Skip,
        UnmappedMemberHandling = JsonUnmappedMemberHandling.Skip,
    };

    /// <summary>
    /// Tries the Firely Terminal cache layout (<c>&lt;dir&gt;/&lt;name&gt;#&lt;version&gt;/package</c>)
    /// first, then falls back to the flat npm layout (<c>&lt;dir&gt;/&lt;name&gt;</c>).
    /// </summary>
    public string ResolvePackageContentDir(
        string fhirPackagesDir,
        string packageName,
        string packageVersion
    )
    {
        var firelyLayout = Path.Combine(
            fhirPackagesDir,
            $"{packageName}#{packageVersion}",
            "package"
        );
        if (Directory.Exists(firelyLayout))
            return firelyLayout;
        return Path.Combine(fhirPackagesDir, packageName);
    }

    public IgPackageModel Scan(string packageContentDir, string packageName, string packageVersion)
    {
        var resources = ReadResources(packageContentDir);

        var codeSystems = new SortedDictionary<string, string>(StringComparer.Ordinal);
        var profiles = new SortedDictionary<string, string>(StringComparer.Ordinal);
        var extensions = new SortedDictionary<string, string>(StringComparer.Ordinal);
        var codeSystemConcepts = new Dictionary<string, IReadOnlyList<ConceptConstant>>();
        var extensionValueTypes = new Dictionary<string, ExtensionValueType>();

        foreach (var resource in resources)
        {
            if (resource.ResourceType != "CodeSystem")
                continue;
            if (resource.Id is null || resource.Url is null)
                continue;
            ClassifyCodeSystem(
                resource,
                NameUtils.ToConstantName(resource.Id),
                resource.Url,
                codeSystems,
                codeSystemConcepts
            );
        }

        var codeSystemUrlByValueSetUrl = IndexSingleCodeSystemValueSets(resources);

        foreach (var resource in resources)
        {
            if (resource.ResourceType != "StructureDefinition")
                continue;
            if (resource.Id is null || resource.Url is null)
                continue;
            ClassifyStructureDefinition(
                resource,
                NameUtils.ToConstantName(resource.Id),
                resource.Url,
                profiles,
                extensions,
                extensionValueTypes,
                codeSystemUrlByValueSetUrl
            );
        }

        var namingSystems = new SortedDictionary<string, NamingSystemEntry>(StringComparer.Ordinal);
        foreach (var resource in resources)
        {
            if (resource.ResourceType != "NamingSystem")
                continue;
            if (resource.Id is null || resource.UniqueId is null || resource.UniqueId.Count == 0)
                continue;
            var constantName = NameUtils.ToConstantName(resource.Id);
            var byTypeMutable = new SortedDictionary<string, List<string>>(StringComparer.Ordinal);
            foreach (var uid in resource.UniqueId)
            {
                if (uid.Type is null || uid.Value is null)
                    continue;
                if (!byTypeMutable.TryGetValue(uid.Type, out var list))
                    byTypeMutable[uid.Type] = list = [];
                list.Add(uid.Value);
            }
            if (byTypeMutable.Count == 0)
                continue;
            var byType = new SortedDictionary<string, IReadOnlyList<string>>(
                StringComparer.Ordinal
            );
            foreach (var (type, values) in byTypeMutable)
                byType[type] = values;
            namingSystems[constantName] = new NamingSystemEntry(resource.Description, byType);
        }

        return new IgPackageModel(
            packageName,
            packageVersion,
            codeSystems,
            profiles,
            extensions,
            codeSystemConcepts,
            extensionValueTypes,
            namingSystems
        );
    }

    private static List<FhirResourceSummary> ReadResources(string packageContentDir)
    {
        var results = new List<FhirResourceSummary>();
        foreach (var file in Directory.EnumerateFiles(packageContentDir, "*.json"))
        {
            using var stream = File.OpenRead(file);
            var resource = JsonSerializer.Deserialize<FhirResourceSummary>(stream, JsonOptions);
            if (resource is not null)
                results.Add(resource);
        }
        return results;
    }

    private static void ClassifyStructureDefinition(
        FhirResourceSummary resource,
        string constantName,
        string url,
        SortedDictionary<string, string> profiles,
        SortedDictionary<string, string> extensions,
        Dictionary<string, ExtensionValueType> extensionValueTypes,
        Dictionary<string, string> codeSystemUrlByValueSetUrl
    )
    {
        if (resource.Kind == "logical" || resource.Derivation == "specialization")
            return;

        if (
            resource.Kind == "complex-type"
            && resource.Derivation == "constraint"
            && resource.Type == "Extension"
        )
        {
            extensions[constantName] = url;
            extensionValueTypes[constantName] = ExtensionValueTypeFor(
                resource,
                codeSystemUrlByValueSetUrl
            );
            return;
        }

        if (resource.Kind == "resource" && resource.Derivation == "constraint")
        {
            profiles[constantName] = resource.Version is null ? url : $"{url}|{resource.Version}";
        }
    }

    private static ExtensionValueType ExtensionValueTypeFor(
        FhirResourceSummary resource,
        Dictionary<string, string> codeSystemUrlByValueSetUrl
    )
    {
        var elements = resource.Snapshot?.Element;
        if (elements is null)
            return ExtensionValueType.None;

        foreach (var element in elements)
        {
            if (element.Path != "Extension.value[x]")
                continue;
            var types = element.Type;
            if (element.Max == "0" || types is null || types.Count == 0)
                return ExtensionValueType.None;
            if (types.Count > 1)
                return ExtensionValueType.ChoiceType;
            var code = types[0].Code;
            if (code is null)
                return ExtensionValueType.None;
            var boundUrl = BoundCodeSystemUrl(elements, element, code, codeSystemUrlByValueSetUrl);
            return boundUrl is null
                ? ExtensionValueType.Fixed(code)
                : ExtensionValueType.BoundCoding(code, boundUrl);
        }
        return ExtensionValueType.None;
    }

    private static string? BoundCodeSystemUrl(
        List<FhirElement> elements,
        FhirElement valueElement,
        string valueTypeCode,
        Dictionary<string, string> codeSystemUrlByValueSetUrl
    )
    {
        if (valueTypeCode != "CodeableConcept" && valueTypeCode != "Coding")
            return null;

        var systemPath =
            valueTypeCode == "CodeableConcept"
                ? "Extension.value[x].coding.system"
                : "Extension.value[x].system";

        foreach (var element in elements)
        {
            if (element.Path == systemPath && element.FixedUri is not null)
                return element.FixedUri;
        }

        var binding = valueElement.Binding;
        if (binding is null || binding.Strength != "required" || binding.ValueSet is null)
            return null;

        var valueSetUrl = binding.ValueSet;
        var sep = valueSetUrl.IndexOf('|');
        var bareUrl = sep < 0 ? valueSetUrl : valueSetUrl[..sep];
        return codeSystemUrlByValueSetUrl.GetValueOrDefault(bareUrl);
    }

    private static Dictionary<string, string> IndexSingleCodeSystemValueSets(
        List<FhirResourceSummary> resources
    )
    {
        var result = new Dictionary<string, string>();
        foreach (var resource in resources)
        {
            if (resource.ResourceType != "ValueSet")
                continue;
            if (resource.Url is null)
                continue;
            var csUrl = SingleCodeSystemUrl(resource.Compose);
            if (csUrl is not null)
                result[resource.Url] = csUrl;
        }
        return result;
    }

    private static string? SingleCodeSystemUrl(FhirCompose? compose)
    {
        if (compose is null)
            return null;
        if (compose.Exclude is { Count: > 0 })
            return null;
        var includes = compose.Include;
        if (includes is null || includes.Count != 1)
            return null;
        FhirComposeInclude only = includes[0];
        if (only.ValueSet is { Count: > 0 })
            return null;
        return only.System;
    }

    private static void ClassifyCodeSystem(
        FhirResourceSummary resource,
        string constantName,
        string url,
        SortedDictionary<string, string> codeSystems,
        Dictionary<string, IReadOnlyList<ConceptConstant>> codeSystemConcepts
    )
    {
        codeSystems[constantName] = url;
        if (resource.Content != "complete" || resource.ConceptList is null)
            return;
        IReadOnlyList<ConceptConstant> concepts = FlattenConcepts(resource.ConceptList);
        if (concepts.Count > 0)
            codeSystemConcepts[constantName] = concepts;
    }

    private static IReadOnlyList<ConceptConstant> FlattenConcepts(List<FhirConcept> concepts)
    {
        var result = new List<ConceptConstant>();
        var usedNames = new HashSet<string>(StringComparer.Ordinal);
        FlattenConcepts(concepts, result, usedNames);
        return result;
    }

    private static void FlattenConcepts(
        List<FhirConcept> concepts,
        List<ConceptConstant> result,
        HashSet<string> usedNames
    )
    {
        foreach (var concept in concepts)
        {
            if (concept.Code is not null)
                result.Add(
                    new ConceptConstant(
                        UniqueMemberName(concept.Code, usedNames),
                        concept.Code,
                        concept.Display
                    )
                );
            if (concept.Children is not null)
                FlattenConcepts(concept.Children, result, usedNames);
        }
    }

    private static string UniqueMemberName(string code, HashSet<string> usedNames)
    {
        var baseName = NameUtils.ToEnumMemberName(code);
        var name = baseName;
        var suffix = 2;
        while (!usedNames.Add(name))
        {
            name = $"{baseName}_{suffix}";
            suffix++;
        }
        return name;
    }
}

File: ig-codegen-py/src/ig_codegen/ig_package_scanner.py
"""Scans a FHIR package directory and classifies its resources into an IgPackageModel."""

from __future__ import annotations

import json
from pathlib import Path

from .ig_package_model import (
    ConceptConstant,
    ExtensionValueType,
    IgPackageModel,
    NamingSystemEntry,
)
from .name_utils import to_constant_name, to_enum_member_name


def resolve_package_content_dir(
    fhir_packages_dir: Path, package_name: str, package_version: str
) -> Path:
    """Returns the package content directory.

    Tries the Firely Terminal cache layout (<dir>/<name>#<version>/package) first,
    then falls back to the npm layout (<dir>/<name>).
    """
    firely = fhir_packages_dir / f"{package_name}#{package_version}" / "package"
    if firely.is_dir():
        return firely
    return fhir_packages_dir / package_name


def scan(package_content_dir: Path, package_name: str, package_version: str) -> IgPackageModel:
    """Reads all JSON files in the package directory and classifies them."""
    resources = _read_resources(package_content_dir)

    code_systems: dict[str, str] = {}
    code_system_concepts: dict[str, list[ConceptConstant]] = {}
    profiles: dict[str, str] = {}
    extensions: dict[str, str] = {}
    extension_value_types: dict[str, ExtensionValueType] = {}

    for resource in resources:
        if resource.get("resourceType") != "CodeSystem":
            continue
        rid = resource.get("id")
        url = resource.get("url")
        if not rid or not url:
            continue
        _classify_code_system(
            resource, to_constant_name(rid), url, code_systems, code_system_concepts
        )

    cs_url_by_valueset_url = _index_single_codesystem_valuesets(resources)

    for resource in resources:
        if resource.get("resourceType") != "StructureDefinition":
            continue
        rid = resource.get("id")
        url = resource.get("url")
        if not rid or not url:
            continue
        _classify_structure_definition(
            resource,
            to_constant_name(rid),
            url,
            profiles,
            extensions,
            extension_value_types,
            cs_url_by_valueset_url,
        )

    naming_systems: dict[str, NamingSystemEntry] = {}
    for resource in resources:
        if resource.get("resourceType") != "NamingSystem":
            continue
        rid = resource.get("id")
        unique_ids = resource.get("uniqueId")
        if not rid or not unique_ids:
            continue
        by_type: dict[str, list[str]] = {}
        for uid in unique_ids:
            uid_type = uid.get("type")
            uid_value = uid.get("value")
            if not uid_type or not uid_value:
                continue
            by_type.setdefault(uid_type, []).append(uid_value)
        if not by_type:
            continue
        naming_systems[to_constant_name(rid)] = NamingSystemEntry(
            description=resource.get("description"),
            by_type=dict(sorted(by_type.items())),
        )

    return IgPackageModel(
        package_name=package_name,
        package_version=package_version,
        code_systems=dict(sorted(code_systems.items())),
        profiles=dict(sorted(profiles.items())),
        extensions=dict(sorted(extensions.items())),
        code_system_concepts=code_system_concepts,
        extension_value_types=extension_value_types,
        naming_systems=dict(sorted(naming_systems.items())),
    )


def _read_resources(package_content_dir: Path) -> list[dict]:
    results = []
    for f in package_content_dir.glob("*.json"):
        try:
            with f.open(encoding="utf-8") as fp:
                data = json.load(fp)
            if isinstance(data, dict):
                results.append(data)
        except (json.JSONDecodeError, OSError):
            pass
    return results


def _classify_code_system(
    resource: dict,
    constant_name: str,
    url: str,
    code_systems: dict[str, str],
    code_system_concepts: dict[str, list[ConceptConstant]],
) -> None:
    code_systems[constant_name] = url
    if resource.get("content") != "complete":
        return
    raw_concepts = resource.get("concept")
    if not raw_concepts:
        return
    used_names: set[str] = set()
    concepts = _flatten_concepts(raw_concepts, used_names)
    if concepts:
        code_system_concepts[constant_name] = concepts


def _flatten_concepts(raw: list[dict], used_names: set[str]) -> list[ConceptConstant]:
    result = []
    for item in raw:
        code = item.get("code")
        if code:
            result.append(
                ConceptConstant(
                    constant_name=_unique_member_name(code, used_names),
                    code=code,
                    display=item.get("display"),
                )
            )
        children = item.get("concept")
        if children:
            result.extend(_flatten_concepts(children, used_names))
    return result


def _unique_member_name(code: str, used_names: set[str]) -> str:
    base = to_enum_member_name(code)
    name = base
    suffix = 2
    while name in used_names:
        name = f"{base}_{suffix}"
        suffix += 1
    used_names.add(name)
    return name


def _classify_structure_definition(
    resource: dict,
    constant_name: str,
    url: str,
    profiles: dict[str, str],
    extensions: dict[str, str],
    extension_value_types: dict[str, ExtensionValueType],
    cs_url_by_valueset_url: dict[str, str],
) -> None:
    kind = resource.get("kind")
    derivation = resource.get("derivation")
    if kind == "logical" or derivation == "specialization":
        return
    is_extension = kind == "complex-type" and derivation == "constraint"
    if is_extension and resource.get("type") == "Extension":
        extensions[constant_name] = url
        extension_value_types[constant_name] = _extension_value_type_for(
            resource, cs_url_by_valueset_url
        )
        return
    if kind == "resource" and derivation == "constraint":
        version = resource.get("version")
        profiles[constant_name] = f"{url}|{version}" if version else url


def _extension_value_type_for(
    resource: dict, cs_url_by_valueset_url: dict[str, str]
) -> ExtensionValueType:
    elements: list[dict] = (resource.get("snapshot") or {}).get("element") or []
    for element in elements:
        if element.get("path") != "Extension.value[x]":
            continue
        types: list[dict] = element.get("type") or []
        if element.get("max") == "0" or not types:
            return ExtensionValueType.NONE
        if len(types) > 1:
            return ExtensionValueType.CHOICE
        code = types[0].get("code")
        if not code:
            return ExtensionValueType.NONE
        bound_url = _bound_code_system_url(elements, element, code, cs_url_by_valueset_url)
        if bound_url:
            return ExtensionValueType.bound_coding(code, bound_url)
        return ExtensionValueType.fixed(code)
    return ExtensionValueType.NONE


def _bound_code_system_url(
    elements: list[dict],
    value_element: dict,
    value_type_code: str,
    cs_url_by_valueset_url: dict[str, str],
) -> str | None:
    if value_type_code not in ("CodeableConcept", "Coding"):
        return None
    system_path = (
        "Extension.value[x].coding.system"
        if value_type_code == "CodeableConcept"
        else "Extension.value[x].system"
    )
    for element in elements:
        if element.get("path") == system_path:
            fixed_uri = element.get("fixedUri")
            if fixed_uri:
                return fixed_uri
    binding: dict = value_element.get("binding") or {}
    if binding.get("strength") != "required":
        return None
    value_set: str | None = binding.get("valueSet")
    if not value_set:
        return None
    bare_url = value_set.split("|")[0]
    return cs_url_by_valueset_url.get(bare_url)


def _index_single_codesystem_valuesets(resources: list[dict]) -> dict[str, str]:
    result: dict[str, str] = {}
    for resource in resources:
        if resource.get("resourceType") != "ValueSet":
            continue
        url = resource.get("url")
        if not url:
            continue
        cs_url = _single_code_system_url(resource.get("compose") or {})
        if cs_url:
            result[url] = cs_url
    return result


def _single_code_system_url(compose: dict) -> str | None:
    if compose.get("exclude"):
        return None
    includes: list[dict] = compose.get("include") or []
    if len(includes) != 1:
        return None
    only = includes[0]
    if only.get("valueSet"):
        return None
    return only.get("system")
